erdjs/src/walletcore/userWallet.ts

```diff
@@ -0,0 +1,156 @@
+        const derivedKeySecondHalf = derivedKey.slice(16, 32);
+        const cipher = crypto.createCipheriv(CipherAlgorithm, derivedKeyFirstHalf, randomness.iv);
+
+        const text = Buffer.concat([secretKey.valueOf(), secretKey.toPublicKey().valueOf()]);
+        const ciphertext = Buffer.concat([cipher.update(text), cipher.final()]);
+        const mac = crypto.createHmac(DigestAlgorithm, derivedKeySecondHalf).update(ciphertext).digest();
+
+        this.publicKey = secretKey.toPublicKey();
+        this.randomness = randomness;
+        this.ciphertext = ciphertext;
+        this.mac = mac;
+        this.kdfparams = kdParams;
+    }
+
+    /**
+     * Copied from: https://github.com/ElrondNetwork/elrond-core-js/blob/v1.28.0/src/account.js#L42
+     * Notes: adjustements (code refactoring, no change in logic), in terms of: 
+     *  - typing (since this is the TypeScript version)
+     *  - error handling (in line with erdjs's error system)
+     *  - references to crypto functions
+     *  - references to object members
+     * 
+     * From an encrypted keyfile, given the password, loads the secret key and the public key.
+     */
+    static decryptSecretKey(keyFileObject: any, password: string): UserSecretKey {
+        const kdfparams = keyFileObject.crypto.kdfparams;
+        const salt = Buffer.from(kdfparams.salt, "hex");
+        const iv = Buffer.from(keyFileObject.crypto.cipherparams.iv, "hex");
+        const ciphertext = Buffer.from(keyFileObject.crypto.ciphertext, "hex");
+        const derivedKey = UserWallet.generateDerivedKey(Buffer.from(password), salt, kdfparams);
+        const derivedKeyFirstHalf = derivedKey.slice(0, 16);
+        const derivedKeySecondHalf = derivedKey.slice(16, 32);
+        
+        const computedMAC = crypto.createHmac(DigestAlgorithm, derivedKeySecondHalf).update(ciphertext).digest();
+        const actualMAC = keyFileObject.crypto.mac;
+
+        if (computedMAC.toString("hex") !== actualMAC) {
+            throw new errors.ErrWallet("MAC mismatch, possibly wrong password");
+        }
+
+        const decipher = crypto.createDecipheriv(keyFileObject.crypto.cipher, derivedKeyFirstHalf, iv);
+
+        let text = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
+        while (text.length < 32) {
+            let zeroPadding = Buffer.from([0x00]);
+            text = Buffer.concat([zeroPadding, text]);
+        }
+
+        let seed = text.slice(0, 32);
+        return new UserSecretKey(seed);
+    }
+
+    /**
+     * Will take about:
+     *  - 80-90 ms in Node.js, on a i3-8100 CPU @ 3.60GHz
+     *  - 350-360 ms in browser (Firefox), on a i3-8100 CPU @ 3.60GHz
+     */
+    private static generateDerivedKey(password: Buffer, salt: Buffer, kdfparams: ScryptKeyDerivationParams): Buffer {
+        // Question for review: @ccorcoveanu, why not this implementation?
+        // https://nodejs.org/api/crypto.html#crypto_crypto_scrypt_password_salt_keylen_options_callback
```

Comment:
If I remember correctly, because scryptsy works in the browser, does the scrypt in the crypto package work also? We can test/benchmark it and if it does, we can switch to it and remove some extra dependencies

Reply:
With the original one, on a i3-8100 CPU @ 3.60GHz:

```
- 80-90 ms in Node.js
- 350-360 ms in browser (Firefox)
```

With the one from the crypto package:
```
 - 15-17 ms in Node.js
 - ... (below)
```

... not added by the `browserify` pipeline: https://www.npmjs.com/package/crypto-browserify. 

Therefore, left as it was.